### ISSUE TITLE
Security of GitOps

### ISSUE THREAD

**Issue by mateuszpruchniak** [open-gitops/documents#57]
Hi Folks 

On the one hand, GitOps increase the security itself through limitation to a minimum of administrative access to infrastructure and configuration, eliminate human error, an audit of any change, declarative single source of truth, automatic verification, and policy enforcement.  Using GitOps tools and technologies, organizations can mitigate the vectors of attack by reducing the number of people and machines that have access to the target system. 
On the other hand, more attention should be paid to securing Git repositories and CI/CD processes.

I would like to open a discussion about the security of GitOps and in particular what are the best security practices. And as a result create a dedicated document.

**Comment by chris-short** [comment]
I welcome this conversation. I would encourage linking to known good guides that already exist as appropriate. I'd also like to see suggestions of what we can do in the GitOps pipeline as well.

**Comment by mateuszpruchniak** [comment]
I will start :)

Basic security setup:
1. Add a policy to enforce signature verification to prevent from applying unauthorized changes on the cluster:
- https://fluxcd.io/legacy/flux/references/git-gpg/
- https://argo-cd.readthedocs.io/en/stable/user-guide/gpg-verification/
2. Enforce policy to block sensitive data being pushed to Git. In order to store secrets and credentials safely in a public or private Git repository, use Mozilla SOPS CLI or SealedSecrets to protect it.
- https://github.com/mozilla/sops
- https://github.com/bitnami-labs/sealed-secrets
3. Rotate SSH keys and Personal Access Tokens, guard against is unauthorized access to Git repositories
4. Enable two-factor authentication for all git users
5. Inclusion strategy in .gitignore file rather than exclusion
6. Hardening of GitOps process 
7. Define access roles on a per-repository basis to ensure only users with valid access credentials are allowed to interact with the repository


of course, not to be missed:
https://www.weave.works/blog/securing-gitops-pipelines

**Comment by chris-short** [comment]
Strong opinions, loosely held.  
  
On point 1:  
We might need to find a new Flux reference. In general, I think this is less important for some environments than others. I know gpg hinders adoption of some tooling too. I always caution involving gpg as a rule so this is me doing that. Yes, my commits are signed, but there are a lot of folks writing code today that don't understand the benefits of it. I'd say we need to point to something or explicitly state the benefits of signing commits.

On point 2:  
I'd like to note there's some serious and substantial movement toward cosign and Chainguard's tooling. If there's a case we could highlight there, it'd be wise even if it was in its infancy, in my opinion given the reception Chainguard/cosign has been given.

On point 3:  
This is good GitHub/GitLab hygiene but is often not done. I wonder if we should have a standards doc or at least point to one about proper care and feeding of GitHub and GitLab instances and not necessarily include specific things in our docs that are good hygiene. I don't want folks to only adopt what we say while missing other important things to do. See point 6.  
  
On point 4:  
See point 3.

On point 5:

> Inclusion strategy in .gitignore file rather than exclusion

Never thought about it like that. Tell me MORE!

On point 6:  
Perfect. We should point to this first maybe and link/highlight to what specifically you MUST do for GitOps security.

On point 9:  
We're going to have to point to something on this one.

On point 10:  
We should be pointing to one of our yet to be created docs for all this maybe?

**Comment by mateuszpruchniak** [comment]
Awesome!
Maybe we can start cooperating on a doc, what do you prefer HackMD or Google Docs?

**Comment by pjbgf** [comment]
Yep completely agree. I think Google docs would make it easier and accessible for most folks keen on collaborating.

**Comment by mateuszpruchniak** [comment]
@pjbgf  @chris-short 

I am not completely proficient in Google Docs, but I created docs based on the longest message :) 

https://docs.google.com/document/d/1aqYabvs1-gurX1lYB5fzQDdptuA7j-LavGCSmxOFM_g

Please, ask for access and I will add anyone interested to contribute to the document.
